### PR TITLE
sql: Allowed columns and indexes to be added and dropped in same transaction

### LINKED ISSUE #19210: sql: Cannot add and then drop a column within the same transaction
You cannot `add` a column and then `drop` it within the scope of the same `transaction` in cockroachdb sql.  When I try this I get an error on the drop statement that the column I am trying to `drop` is in the middle of being added.

Obviously this is ostensibly an odd thing to need to do but the use case is actually pretty common - we have a database migrations project which loops through sql scripts running those which haven't yet been applied to the database. If you run the migrator somewhere where the `add` statement hasn't been applied yet ( build server/fresh deployment ) then the two statements will come together in a single transaction and the migration will fail.

### PR DESCRIPTION
**Pull request by jrote1**
fixes #19210

### CODE REVIEW

**Comment by vivekmenezes** [comment]
I can review this

**Comment by jrote1** [comment]
Have seen the failing test fixing now

**Comment by jrote1** [comment]
@vivekmenezes is all good now

File: pkg/sql/logictest/testdata/logic_test/alter_table [hunk 1]
```diff
@@ -663,3 +663,23 @@ query I
 SELECT COUNT(*) FROM crdb_internal.jobs WHERE status = 'pending' OR status = 'started'
 ----
 0
+
+# Can add and drop column in same transaction
+statement ok
+BEGIN; CREATE TABLE add_drop_column (id INT); ALTER TABLE add_drop_column ADD COLUMN column_to_drop INT; ALTER TABLE add_drop_column DROP COLUMN column_to_drop; COMMIT;
```
Comment (vivekmenezes) [inline pkg/sql/logictest/testdata/logic_test/alter_table]: I'd use table "t" for both these tests

File: pkg/sql/drop.go [hunk 2]
```diff
@@ -427,6 +432,16 @@ func (p *planner) dropIndexByName(
 	return nil
 }
 
+func (p *planner) removeSchemaChangerForMutation(mutationID sqlbase.MutationID) {
```
Comment (vivekmenezes) [inline pkg/sql/drop.go]: You need to find whether the column or indexes is being added through one of the outstanding schema changers. This method would then remove that schema changer and return true

File: pkg/sql/alter_table.go [hunk 3]
```diff
@@ -306,7 +306,12 @@ func (n *alterTableNode) Start(params runParams) error {
 				}
 			}
 			if !found {
-				return fmt.Errorf("column %q in the middle of being added, try again later", t.Column)
+				for _, mutation := range n.tableDesc.Mutations {
+					if mutation.GetColumn().ID == col.ID {
```
Comment (vivekmenezes) [inline pkg/sql/alter_table.go]: the column needs to be a mutation and the mutationID must be part of p.session.TxnState.schemaChangers.schemaChangers (see my comment below)

Once that's true you can safely remove it from the mutation list through

n.tableDesc.Mutations = append(n.tableDesc.Mutations[:i], n.tableDesc.Mutations[i+1:]...)

File: pkg/sql/drop.go [hunk 4]
```diff
@@ -388,7 +388,12 @@ func (p *planner) dropIndexByName(
 		}
 	}
 	if !found {
-		return fmt.Errorf("index %q in the middle of being added, try again later", idxName)
+		for _, mutation := range tableDesc.Mutations {
+			if mutation.GetIndex().ID == idx.ID {
+				p.removeSchemaChangerForMutation(mutation.MutationID)
```
Comment (vivekmenezes) [inline pkg/sql/drop.go]: same here

File: pkg/sql/logictest/testdata/logic_test/alter_table (hunk 1 above)
Comment (vivekmenezes) [reply pkg/sql/logictest/testdata/logic_test/alter_table]: Also add a schema change that adds more than one column

ALTER TABLE add_drop_column ADD COLUMN column_that_stays INT, column_to_drop INT

Comment (vivekmenezes) [reply pkg/sql/logictest/testdata/logic_test/alter_table]: You also want to test this with a UNIQUE column and ensure that the implicit index is dropped.

**Review by vivekmenezes — COMMENTED** [review]
Thanks for making this change! We really appreciate it

**Comment by jrote1** [comment]
I'll look into this this evening

**Comment by jrote1** [comment]
Sorry have been very busy recently, will get this ready to go ASAP

**Comment by jrote1** [comment]
Am having issues with some tests, but don't have time to investigate currently. Closing PR. May reattempt when have time but don't want to hold anything up

**Comment by vivekmenezes** [comment]
Hey thanks for your efforts! I appreciate you giving it a try :-)

**Pull request closed without merging**